node/service/src/tests.rs

```diff
@@ -0,0 +1,715 @@
+			return Some(highest_approved_block_hash);
+		}
+
+		let mut undisputed_chain = Some(highest_approved_block_hash);
+		let mut block_hash = highest_approved_block_hash;
+		while let Some(block) = self.blocks_by_hash.get(&block_hash) {
+			block_hash = block.hash();
+			if ChainBuilder::GENESIS_HASH == block_hash {
+				return None;
+			}
+			let next = block.parent_hash();
+			if self.disputed_blocks.contains(&block_hash) {
+				undisputed_chain = Some(*next);
+			}
+			if block.number() == &base_blocknumber {
+				return None;
+			}
+			block_hash = *next;
+		}
+		undisputed_chain
+	}
+}
+
+impl HeaderProvider<Block> for TestChainStorage {
+	fn header(&self, hash: Hash) -> sp_blockchain::Result<Option<Header>> {
+		Ok(self.blocks_by_hash.get(&hash).cloned())
+	}
+	fn number(&self, hash: Hash) -> sp_blockchain::Result<Option<BlockNumber>> {
+		self.header(hash).map(|opt| opt.map(|h| h.number))
+	}
+}
+
+impl HeaderProviderProvider<Block> for TestChainStorage {
+	type Provider = Self;
+	fn header_provider(&self) -> &Self {
+		self
+	}
+}
+
+#[derive(Debug, Clone)]
+struct ChainBuilder(pub TestChainStorage);
+
+impl ChainBuilder {
+	const GENESIS_HASH: Hash = Hash::repeat_byte(0xff);
+	const GENESIS_PARENT_HASH: Hash = Hash::repeat_byte(0x00);
+
+	pub fn new() -> Self {
+		let mut builder = Self(TestChainStorage::default());
+		let _ = builder.add_block_inner(Self::GENESIS_HASH, Self::GENESIS_PARENT_HASH, 0);
+		builder
+	}
+
+	pub fn add_block<'a>(&'a mut self, hash: Hash, parent_hash: Hash, number: u32) -> &'a mut Self {
+		assert!(number != 0, "cannot add duplicate genesis block");
+		assert!(hash != Self::GENESIS_HASH, "cannot add block with genesis hash");
+		assert!(parent_hash != Self::GENESIS_PARENT_HASH, "cannot add block with genesis parent hash");
+		assert!(self.0.blocks_by_hash.len() < u8::MAX.into());
+		self.add_block_inner(hash, parent_hash, number)
+	}
+
+	fn add_block_inner<'a>(&'a mut self, hash: Hash, parent_hash: Hash, number: u32) -> &'a mut Self {
+		let header = ChainBuilder::make_header(parent_hash, number);
+		assert!(self.0.blocks_by_hash.insert(hash, header).is_none(), "block with hash {:?} already exists", hash,);
+		self.0.blocks_at_height.entry(number).or_insert_with(Vec::new).push(hash);
+		self
+	}
+
+	pub fn ff_approved(
+		&mut self,
+		branch_tag: u8,
+		parent: Hash,
+		block_number: BlockNumber,
+	) -> Hash {
+		let block = self.ff(branch_tag, parent, block_number);
+		let _ = self.0.approved_blocks.insert(block);
+		block
+	}
+
+	/// Add a relay chain block that contains a disputed parachain block.
+	/// For simplicity this is not modeled explicitly.
+	pub fn ff_disputed(
+		&mut self,
+		branch_tag: u8,
+		parent: Hash,
+		block_number: BlockNumber,
+	) -> Hash {
+		let block = self.ff_approved(branch_tag, parent, block_number);
+		let _ = self.0.disputed_blocks.insert(block);
+		block
+	}
+
+	pub fn ff(&mut self, branch_tag: u8, parent: Hash, block_number: BlockNumber) -> Hash {
```

Comment:
what does `ff` stand for? My initial thought was `fast-forward`

Reply:
That's correct, should be expanded.